rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+since it is fundamentally different from the current stack and cannot guarantee
+code reuse.)
+
+In the meantime, we plan to provide limited device integration support for the
+current TensorFlow stack through
+[Modular TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md).
+We anticipate three major functionalities within a device plugin module:
+
+*   Device registration: Will be addressed in a different RFC.
+*   Device management: The focus of this RFC.
+*   Kernel and op registration and implementation:
+    [RFC Accepted](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md). [C API implemented](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/).
+
+[StreamExecutor](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_pimpl.h;l=73) is TensorFlow's main device manager, responsible for work execution and memory management. It provides a set of methods
+(such as
+[Memcpy](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=240))
+that can be customized for a particular device.
+
+We propose a C API wrapper of a subset of methods in
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+as an ABI-stable way to register a custom StreamExecutor platform.
+
+# User Benefits
+
+A decoupled way to add a new device to TensorFlow.
+
+*   Simpler process: Does not have to add a new build toolchain to TensorFlow
+*   Faster time-to-solution: Does not need code review from the TensorFlow team.
+*   Lower maintenance efforts: Only C-API-related changes could break the
+    integration. Unrelated TensorFlow changes would not break the code.
+       *    The C APIs may be changed during the initial experimental phase based 
+            on developer experience and feedback. When the APIs become more mature, 
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+typedef struct SE_PlatformId {
+ void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+typedef struct SE_TimerFns {
+ const size_t struct_size;
```

Comment:
`const` should be dropped to allow devs to initialize struct_size after instantiation.

Reply:
Fixed by @yisitu's commit.